network/p2p/middleware/middleware.go

```diff
@@ -445,18 +503,48 @@ func (m *Middleware) handleIncomingStream(s libp2pnetwork.Stream) {
 
 		// Note: message fields must not be trusted until explicitly validated
 		var msg message.Message
-
 		// read the next message (blocking call)
 		err = r.ReadMsg(&msg)
-
 		if err != nil {
 			if err == io.EOF {
 				break
 			}
 
-			m.log.Err(err).
-				Str("peer_id", remotePeer.String()).
-				Msg("failed to read message")
+			m.log.Err(err).Msg("failed to read message")
+			return
+		}
+
+		// check if unicast messages have reached rate limit before processing next message
+		if !m.unicastRateLimiters.MessageAllowed(remotePeer) {
```

Comment:
https://github.com/onflow/flow-go/pull/2840/commits/959ace63fe02bef19752eb04d42b3e6eb817b0cb

Reply:
The reason we have to check if the message is allowed after the call to `r.ReadMsg` is because we need to break out of the loop when we reach EOF

```
err = r.ReadMsg(&msg)
if err != nil {
	if err == io.EOF {
                 // we reach here when there is no more data to read and avoid falsely rate limiting a message
		break
	}

	m.log.Err(err).Msg("failed to read message")
	return
}
```

If we don't then we will consume a rate limit token even though there was not an actual message. This will cause us to inadvertently rate limit future messages. I've tested this thoroughly.